File: EFQ_Project_Dashboard/modules/custodian_dashboard.py
import streamlit as st
from datetime import datetime
from services.excel_service import (
    get_incidents,
    get_incident_by_id,
    update_incident,
    get_activities_by_incident,
    log_activity,
    create_resolution,
    update_resolution,
    get_resolution_by_incident,
    generate_resolution_id,
)


def render_custodian_dashboard():
    st.title("🔍 Custodian Dashboard")
    
    # KPI Cards
    incidents_df = get_incidents()
    if not incidents_df.empty:
        open_count = len(incidents_df[incidents_df["Status"] != "Closed"])
        closed_count = len(incidents_df[incidents_df["Status"] == "Closed"])
        critical_count = len(incidents_df[incidents_df["Severity"] == "Critical"])
        
        col1, col2, col3, col4 = st.columns(4)
        with col1:
            st.metric("My Open Incidents", open_count)
        with col2:
            st.metric("My Closed Incidents", closed_count)
        with col3:
            st.metric("Critical Items", critical_count)
        with col4:
            st.metric("Total Assigned", len(incidents_df))
    
    st.divider()
    
    # Incident List
    st.subheader("📋 Select Incident to Investigate")
    
    if incidents_df.empty:
        st.info("No incidents assigned to you.")
        return
    
    incident_options = [
        f"[{row['Status']}] {row['IncidentID']} - {row['Severity']} - {row['OEM']}"
        for _, row in incidents_df.iterrows()
    ]
    
    selected_option = st.selectbox("Select incident:", incident_options)
    if not selected_option:
        return
    
    incident_id = selected_option.split(" ")[1]
    incident = get_incident_by_id(incident_id)
    if incident is None:
        st.error("Incident not found.")
        return
    
    # Display Incident Details
    st.subheader(f"Incident: {incident_id}")
    col1, col2, col3, col4 = st.columns(4)
    with col1:
        st.metric("OEM", incident["OEM"])
    with col2:
        st.metric("Severity", incident["Severity"])
    with col3:
        st.metric("Status", incident["Status"])
    with col4:
        st.metric("Date", str(incident["Date"]))
    
    st.write(f"**Customer Complaint:** {incident['CustomerComplaint']}")
    st.write(f"**Description:** {incident['Description']}")
    
    st.divider()
    
    # Investigation Form
    st.subheader("🔬 Investigation & Solution")
    
    resolution = get_resolution_by_incident(incident_id)
    is_new_resolution = resolution is None
    
    if resolution is None:
        resolution = {
            "ResolutionID": generate_resolution_id(),
            "IncidentID": incident_id,
            "RootCause": "",
            "InvestigationDetails": "",
            "Recommendation": "",
            "ProposedSolution": "",
            "CorrectiveAction": "",
            "PreventiveAction": "",
            "ValidationMethod": "",
            "ValidationResult": "Not Tested",
            "ResolutionOwner": "",
            "TargetDate": None,
            "ResolutionDate": None,
            "ResolutionStatus": "Not Started",
            "Remarks": "",
        }
    
    with st.form(f"investigation_form_{incident_id}", clear_on_submit=False):
        col1, col2 = st.columns(2)
        
        with col1:
            st.subheader("Investigation Phase")
            investigation_details = st.text_area(
                "Investigation Details",
                value=resolution.get("InvestigationDetails", ""),
                height=100,
                key=f"investigation_{incident_id}",
            )
            
            root_cause = st.text_area(
                "Root Cause",
                value=resolution.get("RootCause", ""),
                height=100,
                key=f"root_cause_{incident_id}",
            )
        
        with col2:
            st.subheader("Solution Phase")
            recommendation = st.text_area(
                "Recommendation",
                value=resolution.get("Recommendation", ""),
                height=100,
                key=f"rec_{incident_id}",
            )
            
            proposed_solution = st.text_area(
                "Proposed Solution",
                value=resolution.get("ProposedSolution", ""),
                height=100,
                key=f"proposed_{incident_id}",
            )
        
        st.write("---")
        
        # Action Phase
        st.subheader("Action Phase")
        col1, col2 = st.columns(2)
        
        with col1:
            corrective_action = st.text_area(
                "Corrective Action",
                value=resolution.get("CorrectiveAction", ""),
                height=100,
                key=f"corrective_{incident_id}",
            )
        
        with col2:
            preventive_action = st.text_area(
                "Preventive Action",
                value=resolution.get("PreventiveAction", ""),
                height=100,
                key=f"preventive_{incident_id}",
            )
        
        st.write("---")
        
        # Validation Phase
        st.subheader("Validation Phase")
        col1, col2 = st.columns(2)
        
        with col1:
            validation_method = st.text_area(
                "Validation Method",
                value=resolution.get("ValidationMethod", ""),
                height=80,
                key=f"val_method_{incident_id}",
            )
            
            validation_result = st.selectbox(
                "Validation Result",
                ["Not Tested", "Pass", "Fail", "Conditional Pass"],
                index=["Not Tested", "Pass", "Fail", "Conditional Pass"].index(
                    resolution.get("ValidationResult", "Not Tested")
                ),
                key=f"val_result_{incident_id}",
            )
        
        with col2:
            resolution_owner = st.text_input(
                "Resolution Owner",
                value=resolution.get("ResolutionOwner", ""),
                key=f"owner_{incident_id}",
            )
            
            target_date = st.date_input(
                "Target Date",
                value=resolution.get("TargetDate"),
                key=f"target_{incident_id}",
            )
            
            resolution_status = st.selectbox(
                "Resolution Status",
                ["Not Started", "In Progress", "Ready for Validation", "Validated", "Closed"],
                index=["Not Started", "In Progress", "Ready for Validation", "Validated", "Closed"].index(
                    resolution.get("ResolutionStatus", "Not Started")
                ),
                key=f"res_status_{incident_id}",
            )
        
        st.write("---")
        
        remarks = st.text_area(
            "Remarks",
            value=resolution.get("Remarks", ""),
            height=80,
            key=f"remarks_{incident_id}",
        )
        
        # Status Transition Logic
        st.subheader("Status Update")
        col1, col2 = st.columns([0.7, 0.3])
        
        with col1:
            new_status = st.selectbox(
                "Change Incident Status",
                ["New", "Investigation Started", "Root Cause Identified", "Action In Progress", "Validation", "Closed"],
                index=["New", "Investigation Started", "Root Cause Identified", "Action In Progress", "Validation", "Closed"].index(
                    incident.get("Status", "New")
                ),
                key=f"incident_status_{incident_id}",
            )
        
        with col2:
            close_incident = st.checkbox("Mark as Closed", key=f"close_{incident_id}")
        
        # Submit Button
        col1, col2 = st.columns(2)
        with col1:
            submitted = st.form_submit_button("💾 Save Investigation & Solution")
        with col2:
            reset = st.form_submit_button("Reset")
        
        if reset:
            st.rerun()
        
        if submitted:
            resolution_data = {
                "RootCause": root_cause,
                "InvestigationDetails": investigation_details,
                "Recommendation": recommendation,
                "ProposedSolution": proposed_solution,
                "CorrectiveAction": corrective_action,
                "PreventiveAction": preventive_action,
                "ValidationMethod": validation_method,
                "ValidationResult": validation_result,
                "ResolutionOwner": resolution_owner,
                "TargetDate": target_date,
                "ResolutionStatus": resolution_status,
                "Remarks": remarks,
            }
            
            # Save resolution
            if is_new_resolution:
                create_resolution(incident_id, resolution_data)
                log_activity(incident_id, "Resolution Created", st.session_state.get("username", "Unknown"))
            else:
                update_resolution(incident_id, resolution_data)
                log_activity(incident_id, "Resolution Updated", st.session_state.get("username", "Unknown"))
            
            # Update incident status
            if close_incident and new_status == "Closed":
                incident["Status"] = "Closed"
                incident["ResolutionDate"] = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
                log_activity(incident_id, "Status Changed to Closed", st.session_state.get("username", "Unknown"))
            else:
                incident["Status"] = new_status
                log_activity(incident_id, f"Status Changed to {new_status}", st.session_state.get("username", "Unknown"))
            
            update_incident(incident_id, incident)
            st.success("✓ Investigation & solution saved successfully.")
            st.rerun()
    
    # Activity History Expander
    st.write("---")
    with st.expander("📅 Activity History"):
        activities = get_activities_by_incident(incident_id)
        if activities.empty:
            st.info("No activities recorded yet.")
        else:
            for _, activity in activities.iterrows():
                st.write(f"**{activity['Timestamp']}** - {activity['Action']}")
                st.caption(f"User: {activity['User']}")
                st.write("")

File: EFQ_Project_Dashboard/frontend/js/custodian-dashboard.js
document.addEventListener('DOMContentLoaded', async () => {
  const user = await window.EFQNavigation.bootProtectedPage(
    'custodian-dashboard',
    'Custodian Dashboard',
    'Review assigned incidents, capture investigation detail, and close validated actions.'
  );
  if (!user) return;

  const tableBody = document.getElementById('custodianIncidentRows');
  const metrics = document.getElementById('custodianMetrics');
  const detailPanel = document.getElementById('custodianDetailPanel');
  const emptyPanel = document.getElementById('custodianEmptyState');
  const filtersForm = document.getElementById('custodianFilters');
  let incidents = [];
  let currentIncidentId = null;

  async function loadIncidents() {
    try {
      incidents = await window.EFQApi.get('/api/custodian/incidents');
      renderMetrics();
      renderTable();
      if (!incidents.length) {
        emptyPanel.hidden = false;
        detailPanel.hidden = true;
      } else if (!currentIncidentId) {
        await loadDetail(incidents[0].incident_id);
      }
    } catch (error) {
      window.EFQUI.showToast(error.message, 'error');
    }
  }

  function applyFilters() {
    const formValues = Object.fromEntries(new FormData(filtersForm).entries());
    const search = String(formValues.search || '').toLowerCase();
    return incidents.filter((incident) => {
      const matchesSearch = !search || [incident.incident_id, incident.oem, incident.customer_complaint, incident.vehicle_variant, incident.ecu_part_number].join(' ').toLowerCase().includes(search);
      const matchesSeverity = !formValues.severity || incident.severity === formValues.severity;
      const matchesStatus = !formValues.status || incident.status === formValues.status;
      const matchesOem = !formValues.oem || incident.oem === formValues.oem;
      return matchesSearch && matchesSeverity && matchesStatus && matchesOem;
    });
  }

  function renderMetrics() {
    const today = new Date();
    const source = applyFilters();
    const overdue = source.filter((incident) => incident.target_date && new Date(incident.target_date) < today && incident.status !== 'Closed').length;
    const summary = [
      ['My Total Incidents', source.length],
      ['Open Incidents', source.filter((item) => item.status !== 'Closed').length],
      ['Critical Incidents', source.filter((item) => item.severity === 'Critical').length],
      ['Pending Validation', source.filter((item) => item.status === 'Validation').length],
      ['Closed Incidents', source.filter((item) => item.status === 'Closed').length],
      ['Overdue Incidents', overdue],
    ];
    metrics.innerHTML = summary.map(([label, value]) => `<div class="metric-card"><span>${label}</span><strong>${value}</strong></div>`).join('');
  }

  function renderTable() {
    const rows = applyFilters();
    if (!rows.length) {
      tableBody.innerHTML = '<tr><td colspan="8"><div class="empty-state">No incidents are currently assigned to you.</div></td></tr>';
      return;
    }
    tableBody.innerHTML = rows.map((incident) => `
      <tr data-id="${incident.incident_id}">
        <td>${window.EFQUI.escapeHtml(incident.incident_id)}</td>
        <td>${window.EFQUI.escapeHtml(incident.oem)}</td>
        <td>${window.EFQUI.escapeHtml(incident.severity)}</td>
        <td>${window.EFQUI.escapeHtml(incident.issue_type)}</td>
        <td>${window.EFQUI.escapeHtml(incident.customer_complaint)}</td>
        <td>${window.EFQUI.escapeHtml(incident.status)}</td>
        <td>${window.EFQUI.escapeHtml(window.EFQUI.formatDate(incident.target_date))}</td>
        <td>${window.EFQUI.escapeHtml(window.EFQUI.formatDateTime(incident.updated_at))}</td>
      </tr>
    `).join('');
    tableBody.querySelectorAll('tr[data-id]').forEach((row) => row.addEventListener('click', () => loadDetail(row.dataset.id)));
  }

  async function loadDetail(incidentId) {
    currentIncidentId = incidentId;
    const detail = await window.EFQApi.get(`/api/custodian/incidents/${incidentId}`);
    const incident = detail.incident;
    const resolution = detail.resolution || {};
    emptyPanel.hidden = true;
    detailPanel.hidden = false;
    detailPanel.innerHTML = `
      <div class="card">
        <div class="section-head">
          <div>
            <div class="page-title">${window.EFQUI.escapeHtml(incident.incident_id)}</div>
            <div class="muted">${window.EFQUI.escapeHtml(incident.status)} · ${window.EFQUI.escapeHtml(incident.oem)} · ${window.EFQUI.escapeHtml(incident.severity)}</div>
          </div>
          <span class="status-badge">Assigned to ${window.EFQUI.escapeHtml(incident.custodian_name)}</span>
        </div>
      </div>
      <div class="detail-layout">
        <div class="detail-stack">
          <div class="card">
            <h3 class="card-title">Original Incident Information</h3>
            <div class="info-grid">
              <div><strong>Customer Complaint</strong><p>${window.EFQUI.escapeHtml(incident.customer_complaint)}</p></div>
              <div><strong>Issue Type</strong><p>${window.EFQUI.escapeHtml(incident.issue_type)}</p></div>
              <div><strong>Dealer</strong><p>${window.EFQUI.escapeHtml(`${incident.dealer_name}, ${incident.dealer_location}`)}</p></div>
              <div><strong>Dealer Contact</strong><p>${window.EFQUI.escapeHtml(incident.dealer_contact)}</p></div>
              <div><strong>Vehicle</strong><p>${window.EFQUI.escapeHtml(`${incident.vehicle_model} / ${incident.vehicle_variant}`)}</p></div>
              <div><strong>ECU</strong><p>${window.EFQUI.escapeHtml(`${incident.ecu_name} (${incident.ecu_part_number})`)}</p></div>
            </div>
            <p><strong>Description</strong></p>
            <p>${window.EFQUI.escapeHtml(incident.description)}</p>
          </div>
          <div class="card">
            <h3 class="card-title">Activity History</h3>
            <div class="activity-list">
              ${(detail.activities || []).map((activity) => `
                <div class="activity-item">
                  <strong>${window.EFQUI.escapeHtml(activity.action)}</strong>
                  <div class="muted">${window.EFQUI.escapeHtml(activity.user_name)} · ${window.EFQUI.escapeHtml(window.EFQUI.formatDateTime(activity.timestamp))}</div>
                </div>
              `).join('') || '<div class="empty-state">No activity history recorded yet.</div>'}
            </div>
          </div>
        </div>
        <div class="card">
          <h3 class="card-title">Investigation & Solution</h3>
          <form id="resolutionForm" class="grid-form">
            <div class="field"><label for="investigation_details">Investigation Details</label><textarea id="investigation_details" name="investigation_details">${window.EFQUI.escapeHtml(resolution.investigation_details || '')}</textarea></div>
            <div class="field"><label for="root_cause">Root Cause</label><textarea id="root_cause" name="root_cause">${window.EFQUI.escapeHtml(resolution.root_cause || '')}</textarea></div>
            <div class="field"><label for="recommendation">Recommendation</label><textarea id="recommendation" name="recommendation">${window.EFQUI.escapeHtml(resolution.recommendation || '')}</textarea></div>
            <div class="field"><label for="proposed_solution">Proposed Solution</label><textarea id="proposed_solution" name="proposed_solution">${window.EFQUI.escapeHtml(resolution.proposed_solution || '')}</textarea></div>
            <div class="field"><label for="corrective_action">Corrective Action</label><textarea id="corrective_action" name="corrective_action">${window.EFQUI.escapeHtml(resolution.corrective_action || '')}</textarea></div>
            <div class="field"><label for="preventive_action">Preventive Action</label><textarea id="preventive_action" name="preventive_action">${window.EFQUI.escapeHtml(resolution.preventive_action || '')}</textarea></div>
            <div class="grid-form two-column">
              <div class="field"><label for="validation_method">Validation Method</label><input id="validation_method" name="validation_method" value="${window.EFQUI.escapeHtml(resolution.validation_method || '')}"></div>
              <div class="field"><label for="validation_result">Validation Result</label><select id="validation_result" name="validation_result"><option ${resolution.validation_result === 'Not Tested' ? 'selected' : ''}>Not Tested</option><option ${resolution.validation_result === 'Pass' ? 'selected' : ''}>Pass</option><option ${resolution.validation_result === 'Fail' ? 'selected' : ''}>Fail</option><option ${resolution.validation_result === 'Conditional Pass' ? 'selected' : ''}>Conditional Pass</option></select></div>
              <div class="field"><label for="validation_date">Validation Date</label><input id="validation_date" type="date" name="validation_date" value="${window.EFQUI.escapeHtml(resolution.validation_date || '')}"></div>
              <div class="field"><label for="target_date">Target Date</label><input id="target_date" type="date" name="target_date" value="${window.EFQUI.escapeHtml(resolution.target_date || '')}"></div>
              <div class="field"><label for="resolution_status">Resolution Status</label><select id="resolution_status" name="resolution_status"><option ${resolution.resolution_status === 'Not Started' ? 'selected' : ''}>Not Started</option><option ${resolution.resolution_status === 'Investigation' ? 'selected' : ''}>Investigation</option><option ${resolution.resolution_status === 'Root Cause Identified' ? 'selected' : ''}>Root Cause Identified</option><option ${resolution.resolution_status === 'Action In Progress' ? 'selected' : ''}>Action In Progress</option><option ${resolution.resolution_status === 'Validation' ? 'selected' : ''}>Validation</option><option ${resolution.resolution_status === 'Closed' ? 'selected' : ''}>Closed</option></select></div>
            </div>
            <div class="field"><label for="remarks">Remarks</label><textarea id="remarks" name="remarks">${window.EFQUI.escapeHtml(resolution.remarks || '')}</textarea></div>
            <div class="form-actions"><button type="submit" class="primary">Save Investigation &amp; Solution</button></div>
          </form>
        </div>
      </div>
    `;

    document.getElementById('resolutionForm').addEventListener('submit', async (event) => {
      event.preventDefault();
      const button = event.currentTarget.querySelector('button[type="submit"]');
      const payload = Object.fromEntries(new FormData(event.currentTarget).entries());
      window.EFQUI.setButtonLoading(button, true, 'Saving solution...');
      try {
        await window.EFQApi.put(`/api/custodian/incidents/${incidentId}/resolution`, payload);
        window.EFQUI.showToast('Investigation and solution saved successfully.', 'success');
        await loadIncidents();
        await loadDetail(incidentId);
      } catch (error) {
        window.EFQUI.showToast(error.message, 'error');
      } finally {
        window.EFQUI.setButtonLoading(button, false);
      }
    });
  }

  filtersForm.addEventListener('input', () => {
    renderMetrics();
    renderTable();
  });
  await loadIncidents();
});
